### ISSUE TITLE
Make batch parameter support numpy or pil array source type

### ISSUE THREAD

**Issue by darouwan** [ultralytics/ultralytics#24218]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Currently, the `batch` parameter in `model.predict()` only functions when the source is a directory, video file, or .txt file. According to the [Inference Sources documentation](https://docs.ultralytics.com/modes/predict/#inference-sources), these sources allow the internal loader to stream or batch the data.

However, when a user passes a large List of images (e.g., `List[np.ndarray]` or `List[PIL.Image]`), the batch argument is ignored, and the model attempts to process the entire list at once. This frequently leads to Out of Memory (OOM) errors on both GPU and system RAM for high-resolution datasets or long lists.

### Use case

Users often pre-load or pre-process images in memory using OpenCV or PIL before passing them to the model. Without batching support for lists, users are forced to manually chunk their lists into smaller sub-lists, which adds boilerplate code and bypasses the internal efficiency of the Ultralytics predictor.

We can modify the internal predictor logic to respect the batch size when the input source is a list. If batch is specified and the input is a list, the predictor should iterate through the list in chunks of size n.

### Additional

I am happy to help implement this change if the maintainers agree with this approach.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @darouwan, thank you for your interest in Ultralytics 🚀 and for the thoughtful feature request around `batch` support for in-memory `List[np.ndarray]` and `List[PIL.Image]` sources.

This is an automated response to help get your issue triaged quickly 🤖, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Since this appears to be a feature request ✨, it would be especially helpful if you could share any additional implementation details you have in mind, such as expected behavior for chunked iteration, streaming compatibility, and how you’d like results ordering to behave for list inputs. If you move forward with a PR, thank you for being willing to contribute 🙌

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
This looks valid: list/PIL/NumPy inputs currently go through [`LoadPilAndNumpy`](https://docs.ultralytics.com/reference/data/loaders/#ultralytics.data.loaders.LoadPilAndNumpy), which iterates only once, so `batch` is effectively ignored for in-memory sources; a small PR that adds chunked iteration for that path, keeps single-image behavior unchanged, and includes tests plus a short docs update would be welcome.

**Comment by darouwan** [comment]
@glenn-jocher  I have implemented it in https://github.com/ultralytics/ultralytics/pull/24254 . You can also help to test it with new parameter `batch` in model.predict() method.

**Comment by glenn-jocher** [comment]
Thanks for opening `#24254`; please keep the change scoped to the in-memory loader path and reuse the existing `batch` override in `model.predict()` rather than changing its signature, and add tests for `list[np.ndarray]`, `list[PIL.Image]`, single-image input, and non-divisible final batches so we can review it cleanly.

**Comment by darouwan** [comment]
@glenn-jocher I think I am still using in memory loader and reuse the original batch parameter by enabling it in pil and numpy loader. And what do you mean for changing signature?
Do you mean my code will auto enable batch split and cannot run it like before to process entire batch?

**Comment by glenn-jocher** [comment]
By “changing the signature” I meant please don’t add a new named `batch` argument to `Model.predict()` or `Model.__call__`, since `batch` is already passed through `**kwargs`; and no, the goal is not to force splitting for every in-memory source, but to keep current behavior when `batch` is omitted and only chunk `list`/PIL/NumPy inputs when the user explicitly sets `batch` (for example, `batch=8`).

**Comment by darouwan** [comment]
@glenn-jocher  Got it. I don't add new `batch` argument in `model.predict()` method, where just use the existed parameter from `**kwargs`. I am trying to keep the original process way if user doesn't set `batch` explicitly. But it's a little tricky as the `batch` parameter has a default value `1`, do you have good idea to distinguish them? 

My last commit can impletement it here https://github.com/darouwan/ultralytics/commit/4f9849a54c15f60644b78873e8f4011b7869cb90, but I am not sure if it is graceful enough~

**Comment by glenn-jocher** [comment]
Yes, I’d avoid inferring from `batch == 1`; instead capture whether the user explicitly passed `batch` before merging defaults in `Model.predict()` (for example with a private `batch_provided = "batch" in kwargs` flag on the predictor) and only let `LoadPilAndNumpy` chunk list inputs when that flag is `True`, otherwise keep the current one-shot behavior for in-memory sources.

**Comment by darouwan** [comment]
@glenn-jocher  OK~ I think I am implementing in the same way you planed. I use `if_set_batch_explicitly` to indicate if `batch` is set  explicitely from predict() and then pass it to `LoadPilAndNumpy` in `loaders.py` to judge if it should keep current behavior. 

And where should I add tests for `list[np.ndarray]`, `list[PIL.Image]`, single-image input, and non-divisible final batches?  I add new test cases here (https://github.com/ultralytics/ultralytics/pull/24254/changes/04d5e87cfb18926f1fb20d1ceb5c38849a3f8514), is it OK?

**Comment by glenn-jocher** [comment]
Yes, that’s the right direction—please keep these in the existing Python predict/inference tests rather than a new test module, and make sure they assert both unchanged default behavior when `batch` is not explicitly passed and correct chunking for `list[np.ndarray]`, `list[PIL.Image]`, single-image input, and a non-divisible final batch.

**Comment by darouwan** [comment]
@glenn-jocher  OK. I have added all necessary test cases in PR https://github.com/ultralytics/ultralytics/pull/24254 including default bahavior when batch is not explicitly and correct chunking for `list[np.ndarray]`, `list[PIL.Image]`, single-image input, and a non-divisible final batch.

**Comment by glenn-jocher** [comment]
Thanks for updating `#24254`; please keep any follow-up changes and discussion on the PR thread so the review stays in one place.

**Comment by darouwan** [comment]
OK. Hope this PR can be approved soon.

**Comment by glenn-jocher** [comment]
Thanks, we’ll review `#24254` on the PR thread as time allows; there’s no ETA, but keeping any further updates there will help the review.
